src/main/java/bio/terra/service/filedata/azure/util/BlobContainerCopierBuilder.java

```diff
@@ -0,0 +1,141 @@
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceDestinationPairs(
+      List<BlobCopySourceDestinationPair> pairs) {
+    sourceDestinationPairs = pairs;
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceClientFactory(BlobContainerClientFactory factory) {
+    sourceClientFactory = Objects.requireNonNull(factory, "Source client factory can't be null");
+
+    return this;
+  }
+
+  public BlobContainerCopierBuilder destinationClientFactory(BlobContainerClientFactory factory) {
+    destinationClientFactory =
+        Objects.requireNonNull(factory, "Destination client factory can't be null");
+
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceContainerPrefix(String prefix) {
+
+    sourceContainerPrefix = Objects.requireNonNull(prefix, "The prefix can't be null");
+
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceBlobUrl(String url) {
+    BlobUrlParts blobUrl = BlobUrlParts.parse(url);
+
+    if (Strings.isEmpty(blobUrl.getBlobContainerName())) {
+      throw new IllegalArgumentException(
+          appendBlobGuidanceToErrorMessage("Container name is missing."));
+    }
+
+    if (Strings.isEmpty(blobUrl.getBlobName())) {
+      throw new IllegalArgumentException(
+          appendBlobGuidanceToErrorMessage("URL must contain a blob name."));
+    }
+
+    String permissions = blobUrl.getCommonSasQueryParameters().getPermissions();
+
+    if (Strings.isEmpty(permissions) || !permissions.contains("r")) {
+      throw new IllegalArgumentException(
+          appendBlobGuidanceToErrorMessage("Read permission is required."));
+    }
+
+    sourceBlobUrl = url;
+
+    return this;
+  }
+
+  public String appendBlobGuidanceToErrorMessage(String errorMsg) {
+    return new StringBuilder(errorMsg)
+        .append(" The URL format must be:")
+        .append("https://{account name}.blob.core.windows.net/{container name}/{blob name}")
+        .append("?{valid SAS with read permissions}")
+        .toString();
+  }
+
+  public BlobContainerCopierBuilder pollingInterval(Duration pollingInterval) {
+    Objects.requireNonNull(pollingInterval, "Invalid polling interval. It can't be null");
+
+    if (pollingInterval.getSeconds() < MIN_POLLING_INTERVAL) {
+      throw new IllegalArgumentException(
+          "Polling interval can't be less than " + MIN_POLLING_INTERVAL + " seconds");
+    }
+    this.pollingInterval = pollingInterval;
+
+    return this;
+  }
+
+  public BlobContainerCopierBuilder listOperationTimeOut(Duration listOperationTimeout) {
+    Objects.requireNonNull(listOperationTimeout, "Invalid timeout value. It can't be null");
+
+    if (listOperationTimeout.getSeconds() < MIN_LIST_TIMEOUT) {
+      throw new IllegalArgumentException(
+          "List polling timeout can't be less than " + MIN_LIST_TIMEOUT + " seconds");
+    }
+    this.listOperationTimeout = listOperationTimeout;
+
+    return this;
+  }
+
+  public BlobContainerCopier build() {
```

Comment:
I think this API is confusing because it looks like a builder pattern but it's not a builder, since the fields it's setting aren't `final`. The correct way to write a builder is to have the fields be modifiable in the builder object, but final in the built object. This can be done by making all the fields parameters to the class being built.

Reply:
This is related to the same issue raised by @nmalfroy , re: making the source information immutable. This is something we can refactor later and will make the all fields in the copier `final`.